code/over_time.py

```diff
@@ -0,0 +1,152 @@
+"""
+Functions to jump through the git history of a repo.
+"""
+from datetime import datetime
+import pathlib
+import subprocess
+from typing import Callable, Dict, Optional, Sequence, Tuple, TypeVar, Union
+
+import pandas as pd
+
+from code.complexity_metrics import COMPLEXITY_METRICS, get_repo_complexities
+
+try:
+    from fastprogress import progress_bar
+except ModuleNotFoundError:
+    progress_bar = lambda x: None
+
+
+PathLike = Union[str, pathlib.Path]
+T = TypeVar("T")
+
+
+def git_log(dp: PathLike) -> Tuple[str, ...]:
+    """Returns a tuple of all commit hashes in the git history (newest first)."""
+    output = subprocess.check_output(["git", "-C", str(dp), "log", '--format=format:"%H"'])
+    output = output.strip().decode("ascii")
+    output = output.replace('"', "")
+    return tuple(output.split("\n"))
+
+
+def git_commit_timestamps(dp: PathLike) -> Dict[str, datetime]:
+    """Returns a tuple of all commit hashes in the git history (newest first)."""
+    output = subprocess.check_output(["git", "-C", str(dp), "log", '--format=format:"%H|%ci"'])
+    output = output.strip().decode("ascii")
+    output = output.replace('"', "")
+    result = {}
+    for row in output.split("\n"):
+        cid, ts = row.split("|")
+        result[cid] = datetime.fromisoformat(ts)
+    return result
+
+
+def git_status(dp: PathLike) -> str:
+    """Returns the git status message."""
+    output = subprocess.check_output(["git", "-C", str(dp), "status"])
+    output = output.strip().decode("ascii")
+    return output
+
+
+def git_current_branch(dp: PathLike) -> str:
+    """Determines the name of the currently checked-out branch."""
+    status = git_status(dp)
+    return status.split("\n")[0].replace("On branch ", "")
+
+
+def git_checkout(dp: PathLike, commit_or_branch: str):
+    """Check out a specific branch or commit in the repository under `dp`."""
+    output = subprocess.check_output(["git", "-C", str(dp), "checkout", commit_or_branch])
```

Comment:
When I ran this inside a notebook, this outputted a lot of messages 
```
switch to...
You are in 'detached HEAD' state
....
```
I don't quite understand why but it disappeared when I piped stderr to devnull:
```suggestion
    output = subprocess.check_output(["git", "-C", str(dp), "checkout", commit_or_branch], stderr=subprocess.DEVNULL)
```